enhancements/cluster-api/crd-compatibility-checker.md

```diff
@@ -0,0 +1,469 @@
+
+* As a cluster administrator, I want OpenShift to ensure that objects created for OpenShift do not change in behaviour on upgrade.
+
+* As a platform operator, I want to coordinate multiple compatibility requirements from different actors so that all stakeholders can safely use the same CRDs.
+
+### Goals
+
+- Enable OpenShift to safely allow external workloads to manage CRDs that are also used by OpenShift payload components
+- Provide a mechanism for multiple actors to specify compatibility requirements for the same CRD
+- Prevent installation of incompatible CRD versions through admission webhooks
+- Allow controllers to programmatically check compatibility before attempting upgrades
+- Enable object schema validation in specific namespaces to maintain upgradeability
+- Minimize operational changes required for existing external workloads
+
+### Non-Goals
+
+- Defining the specific algorithm for determining CRD compatibility (assumed to be described elsewhere)
+- Enforcing single ownership of CRDs (must be ensured out-of-band)
+- Providing a user interface for managing compatibility requirements
+- Resolving compatibility conflicts between different actors
+
+## Proposal
+
+The proposal introduces a new admission webhook for CRDs.
+We introduce a new cluster-scoped CRD called `CRDCompatibilityRequirement` that allows actors to express compatibility requirements for target CRDs.
+The admission webhook ensures that subsequent changes to referenced CRDs are compatible with all requirements specified in a `CRDCompatibilityRequirement`.
+
+### Workflow description
+
+The following are 2 related example uses of the proposed API and controller.
+
+#### CCAPIO ensures cluster stability while HCP manages CAPI CRDs
+
+In this example, Hosted Control Planes (HCP) takes responsibility for the management of the cluster-api (CAPI) `Cluster` CRD from cluster-capi-operator (CCAPIO).
+
+CCAPIO deploys CAPI version X and uses it to manage cluster resources.
+
+HCP deploys CAPI version Y, where Y>X. HCP uses it to deploy external managed clusters.
+
+As `Cluster` can only have a single CRD in the management cluster, they must both use the same CRD.
+
+* HCP will assert ownership of the `Cluster` CRD
+* CCAPIO will assert its requirements for the `Cluster` CRD, but will no longer manage the CRD.
+* HCP will upgrade the `Cluster` CRD
+
+CRD Compatibility Checker ensures that the version HCP upgrades to does not break CCAPIO, and therefore the management cluster.
+
+During normal operation, CCAPIO applies sets of resources to the cluster which are defined in 'transport' config maps.
+CCAPIO will be enhanced such that it will automatically create a CRDCompatibilityRequirement for any CRD it loads.
+In the case of `Cluster` this would be:
+
+```yaml
+apiVersion: operator.openshift.io/v1alpha1
+kind: CRDCompatibilityRequirement
+metadata:
+  name: cluster-api-cluster-ccapio
+spec:
+  crdRef: clusters.cluster.x-k8s.io
```

Comment:
We avoid using `Ref` in openshift APIs, and these are normally structs with GVR. I guess we don't need the GVR here but perhaps the following would be more in keeping

```
crd:
  name: clusters.cluster.x-k8s.io
```